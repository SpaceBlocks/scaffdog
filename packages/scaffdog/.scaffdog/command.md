---
name: 'command'
root: 'src/commands'
output: '.'
questions:
  name:
    message: 'Please enter command name.'
  summary:
    message: 'Please enter command summary.'
    initial: 'TODO'
---

# Variables

- name: `{{ inputs.name | kebab }}`
- cmd: `{{ inputs.name | camel }}`

# `../commands/index.ts`

```typescript
{{- blank := "^$" -}}
{{- file := read output.abs -}}
{{- header := file | before blank -}}
{{- body := file | after blank -}}

{{ header }}
import {{ cmd }} from './{{ name }}';

{{ body | replace "];" ("," + cmd + "];") }}
```

# `{{ name }}.ts`

```typescript
import { createCommand } from '{{ relative "../src/command" }}';

export default createCommand({
  name: '{{ name }}',
  summary: '{{ inputs.summary }}',
  args: {},
  flags: {},
})(async ({ cwd, logger, args, flags }) => {
  logger.debug(cwd, args, flags);
  throw new Error('unimplemented'); // TODO
});
```

# `{{ name }}.test.ts`

```typescript
import { expect, test } from 'vitest';
import { runCommand } from '{{ relative "../src/mocks/command-test-utils" }}';
import cmd from './{{ name }}';

const defaults = {
  args: {},
  flags: {},
};

test('TODO', async () => {
  const { code } = await runCommand(
    cmd,
    {
      ...defaults.args,
    },
    {
      ...defaults.flags,
    },
  );

  expect(code).toBe(0);
});
```
