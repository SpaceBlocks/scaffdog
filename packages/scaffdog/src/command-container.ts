import type { CommandModule, CommandList } from './command';

export class CommandContainer {
  private _commands: CommandList;

  public constructor(commands: CommandModule<any, any>[]) {
    const map = new Map();

    commands.forEach((module) => {
      map.set(module.name, module);

      if (module.commands != null) {
        module.commands.forEach((child) => {
          map.set(`${module.name}.${child.name}`, child);
        });
      }
    });

    this._commands = map;
  }

  public all(): CommandModule[] {
    return Array.from(this._commands.values());
  }

  public main(): CommandModule[] {
    const commands: CommandModule[] = [];

    this._commands.forEach((module, key) => {
      if (!key.includes('.')) {
        commands.push(module);
      }
    });

    return commands;
  }

  public get(name: string): CommandModule | undefined {
    return this._commands.get(name);
  }

  public mustGet(name: string): CommandModule {
    const module = this.get(name);
    if (module == null) {
      throw new ReferenceError(`"${name}" command does not exists`);
    }

    return module;
  }
}
