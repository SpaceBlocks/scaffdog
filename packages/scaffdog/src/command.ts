import type { Consola } from 'consola';
import type { Merge, PackageJson } from 'type-fest';
import type yargs from 'yargs';
import type { CommandContainer } from './command-container';
import type { globalFlags } from './global-flags';

export type GlobalFlags = yargs.InferredOptionTypes<typeof globalFlags>;
export type GlobalFlagsWith<T> = Merge<GlobalFlags, T>;

export type CommandOption = {
  [key: string]: yargs.Options;
};

export type CommandArgs<T extends CommandOption> = yargs.InferredOptionTypes<T>;

export type CommandContext<A extends CommandOption, F extends CommandOption> = {
  cwd: string;
  pkg: PackageJson;
  container: CommandContainer;
  logger: Consola;
  size: {
    rows: number;
    columns: number;
  };
  args: CommandArgs<A>;
  flags: GlobalFlagsWith<CommandArgs<F>>;
};

export type CommandRunner<A extends CommandOption, F extends CommandOption> = (
  context: CommandContext<A, F>,
) => Promise<number>;

export type CommandDefinition<
  A extends CommandOption = CommandOption,
  F extends CommandOption = CommandOption,
> = {
  name: string;
  summary: string;
  args: A;
  flags: F;
  commands?: CommandModule<any, any>[];
};

export type CommandModule<
  A extends CommandOption = CommandOption,
  F extends CommandOption = CommandOption,
> = CommandDefinition<A, F> & {
  run: CommandRunner<A, F>;
  build: yargs.BuilderCallback<any, any>;
};

export type CommandList = Map<string, CommandModule<any, any>>;

export type CommandFactory<
  A extends CommandOption = CommandOption,
  F extends CommandOption = CommandOption,
> = (runner: CommandRunner<A, F>) => CommandModule<A, F>;

const buildCommandSignature = (name: string, args: CommandOption): string => {
  const arr = [name];

  Object.entries(args).forEach(([key, opts]) => {
    const inner = opts.array ? `${key}..` : key;
    arr.push(opts.demandOption ? `<${inner}>` : `[${inner}]`);
  });

  return arr.join(' ');
};

export const buildCommand = <A extends CommandOption, F extends CommandOption>(
  yargs: yargs.Argv<any>,
  module: CommandModule<A, F>,
): yargs.Argv<any> => {
  return yargs.command(
    buildCommandSignature(module.name, module.args),
    '',
    module.build,
  );
};

export const createCommand = <A extends CommandOption, F extends CommandOption>(
  definition: CommandDefinition<A, F>,
): CommandFactory<A, F> => {
  return (run) => {
    return {
      ...definition,
      build: (yargs) => {
        yargs.options(definition.flags);

        Object.entries(definition.args).forEach(([key, opts]) => {
          yargs.positional(key, opts as any);
        });

        if (definition.commands != null) {
          definition.commands.forEach((module) => {
            buildCommand(yargs, module);
          });
        }

        return yargs;
      },
      run,
    };
  };
};
