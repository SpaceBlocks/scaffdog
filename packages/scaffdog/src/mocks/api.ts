import { vi } from 'vitest';
import type { Scaffdog, ScaffdogInitializer } from '../api';
import { createResolvedConfig } from '../lib/config.factory';

export const createScaffdogMock = (
  mocks: Partial<Scaffdog> = {},
): Scaffdog => ({
  path: {
    project: '',
    config: '',
  },
  config: createResolvedConfig(),
  list: vi.fn(),
  generate: vi.fn(),
  ...mocks,
});

export const createScaffdogInitializerMock = (
  mocks: Partial<ScaffdogInitializer> = {},
): ScaffdogInitializer => {
  const scaffdog = createScaffdogMock();

  return {
    createScaffdog: vi.fn().mockReturnValueOnce(scaffdog),
    loadScaffdog: vi.fn().mockResolvedValueOnce(scaffdog),
    ...mocks,
  };
};
