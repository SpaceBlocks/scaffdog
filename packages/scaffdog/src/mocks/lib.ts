import type { Library } from '../lib';
import { createConfigLibraryMock } from '../lib/config.mock';
import { createDocumentLibraryMock } from '../lib/document.mock';
import { createErrorLibraryMock } from '../lib/error.mock';
import { createFsLibraryMock } from '../lib/fs.mock';
import { createPromptLibraryMock } from '../lib/prompt.mock';
import { createQuestionLibraryMock } from '../lib/question.mock';

export const createLibraryMock = (mocks: Partial<Library> = {}): Library => {
  return {
    fs: createFsLibraryMock(),
    error: createErrorLibraryMock(),
    prompt: createPromptLibraryMock(),
    config: createConfigLibraryMock(),
    question: createQuestionLibraryMock(),
    document: createDocumentLibraryMock(),
    ...mocks,
  };
};
