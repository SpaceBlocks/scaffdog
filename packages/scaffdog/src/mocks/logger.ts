import { PassThrough } from 'stream';
import type { ConsolaInstance } from 'consola';
import { LogLevels, createConsola } from 'consola';
import stripAnsi from 'strip-ansi';

export const createLogger = (): {
  logger: ConsolaInstance;
  getStdout: () => string;
  getStderr: () => string;
} => {
  const output = {
    stdout: '',
    stderr: '',
  };

  const stream = {
    stdout: new PassThrough(),
    stderr: new PassThrough(),
  };

  const logger = createConsola({
    fancy: true,
    formatOptions: {
      date: false,
    },
    level: LogLevels.info,
    ...(stream as any),
  });

  stream.stdout.on('data', (chunk) => (output.stdout += chunk));
  stream.stderr.on('data', (chunk) => (output.stderr += chunk));

  return {
    logger,
    getStdout: () => stripAnsi(output.stdout),
    getStderr: () => stripAnsi(output.stderr),
  };
};
