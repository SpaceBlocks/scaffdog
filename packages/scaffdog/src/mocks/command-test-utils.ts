import path from 'path';
import type {
  AnyCommandModule,
  CommandArgs,
  CommandModule,
  CommandOption,
  GlobalFlags,
  GlobalFlagsWith,
  InferredCommandModuleArgs,
  InferredCommandModuleFlags,
} from '../command';
import type { CommandContainer } from '../command-container';
import { createCommandContainer } from '../command-container';
import type { Library } from '../lib';
import { createLibraryMock } from './lib';
import { createLogger } from './logger';

export const cwd = path.resolve(__dirname, '../../');

export type CommandRunnerParams<T extends AnyCommandModule> = {
  args?: Partial<CommandArgs<InferredCommandModuleArgs<T>>>;
  flags?: Partial<GlobalFlagsWith<CommandArgs<InferredCommandModuleFlags<T>>>>;
  lib?: Library;
  container?: CommandContainer;
};

export type CommandRunner<T extends AnyCommandModule> = (
  params: CommandRunnerParams<T>,
) => Promise<{
  code: number;
  stdout: string;
  stderr: string;
}>;

export const createCommandRunner =
  <A extends CommandOption, F extends CommandOption>(
    cmd: CommandModule<A, F>,
    defaults: {
      args: CommandArgs<A>;
      flags: CommandArgs<F>;
    },
  ): CommandRunner<CommandModule<A, F>> =>
  async ({ args, flags, lib, container }) => {
    const { logger, getStdout, getStderr } = createLogger();

    const globalFlags: GlobalFlags = {
      project: '.scaffdog',
      help: false,
      version: false,
      verbose: true,
    };

    const code = await cmd.run({
      cwd,
      pkg: {
        name: 'scaffdog',
        version: '1.0.0',
        description: 'scaffdog is Markdown driven scaffolding tool.',
      },
      size: {
        rows: 120,
        columns: 60,
      },
      logger,
      container: container ?? createCommandContainer([]),
      lib: lib ?? createLibraryMock({}),
      args: {
        ...defaults.args,
        ...(args ?? {}),
      },
      flags: {
        ...defaults.flags,
        ...globalFlags,
        ...(flags ?? {}),
      } as never,
    });

    return {
      code,
      stdout: getStdout(),
      stderr: getStderr(),
    };
  };
