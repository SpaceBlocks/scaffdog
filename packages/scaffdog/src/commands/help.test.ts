import { expect, test } from 'vitest';
import { createCommand } from '../command';
import { CommandContainer } from '../command-container';
import { runCommand } from '../mocks/command-test-utils';
import cmd from './help';

const defaults = {
  args: {
    command: [],
  },
  flags: {},
};

test('command', async () => {
  const foo = createCommand({
    name: 'foo',
    summary: 'foo summary.',
    args: {
      name: {
        type: 'string',
        description: 'foo.name.description.',
      },
    },
    flags: {
      key1: {
        type: 'string',
        description: 'key1.description.',
      },
      key2: {
        type: 'boolean',
        description: 'key2.description.',
      },
      key3: {
        type: 'number',
        description: 'key3.description.',
        alias: 'k',
      },
    },
  })(async () => 0);

  const container = new CommandContainer([foo]);

  const { code, stdout } = await runCommand(
    cmd,
    {
      ...defaults.args,
      command: ['foo'],
    },
    {
      ...defaults.flags,
    },
    container,
  );

  expect(code).toBe(0);

  expect(stdout).toMatchSnapshot();
});

test('global', async () => {
  const { code, stdout } = await runCommand(cmd, defaults.args, defaults.flags);

  expect(code).toBe(0);

  expect(stdout).toMatchSnapshot();
});
