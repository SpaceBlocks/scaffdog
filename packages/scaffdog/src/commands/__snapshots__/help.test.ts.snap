// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`command 1`] = `
"foo summary.

USAGE
  scaffdog foo <name> [flags]

COMMAND FLAGS
      --key1     key1.description.
      --key2     key2.description.
  -k, --key3     key3.description.

GLOBAL FLAGS
  -p, --project  Directory to load the scaffdog project.
  -v, --verbose  Enable logging.
      --help     Show help.
      --version  Output the version number.
"
`;

exports[`global 1`] = `
"scaffdog is Markdown driven scaffolding tool.

USAGE
  scaffdog <command> [flags]

GLOBAL FLAGS
  -p, --project  Directory to load the scaffdog project.
  -v, --verbose  Enable logging.
      --help     Show help.
      --version  Output the version number.

Run \`scaffdog help [command..]\` for help with a specific command.
"
`;
